idb/postgres/internal/writer/writer.go

```diff
@@ -327,119 +289,67 @@ func (w *Writer) writeAccountDeltas(round basics.Round, deltas ledgercore.Accoun
-				}
+				batch.Queue(deleteAppStmtName, uint64(index), creator[:], uint64(round))
 			}
 		}
 	}
-
-	return nil
 }
 
-func (w *Writer) writeDeletedAssetHoldings(round basics.Round, modifiedAssetHoldings map[ledgercore.AccountAsset]bool) error {
+func writeDeletedAssetHoldings(round basics.Round, modifiedAssetHoldings map[ledgercore.AccountAsset]bool, batch *pgx.Batch) {
 	for aa, created := range modifiedAssetHoldings {
 		if !created {
-			_, err := w.tx.Exec(
-				context.Background(), deleteAccountAssetStmtName,
-				aa.Address[:], uint64(aa.Asset), uint64(round))
-			if err != nil {
-				return fmt.Errorf(
-					"writeDeletedAssetHoldings() exec delete account asset err: %w", err)
-			}
+			address := new(basics.Address)
+			*address = aa.Address
+
+			batch.Queue(
+				deleteAccountAssetStmtName, address[:], uint64(aa.Asset), uint64(round))
 		}
 	}
-
-	return nil
 }
 
-func (w *Writer) writeDeletedAppLocalStates(round basics.Round, modifiedAppLocalStates map[ledgercore.AccountApp]bool) error {
+func writeDeletedAppLocalStates(round basics.Round, modifiedAppLocalStates map[ledgercore.AccountApp]bool, batch *pgx.Batch) {
 	for aa, created := range modifiedAppLocalStates {
 		if !created {
-			_, err := w.tx.Exec(
-				context.Background(), deleteAccountAppStmtName,
-				aa.Address[:], uint64(aa.App), uint64(round))
-			if err != nil {
-				return fmt.Errorf(
-					"writeDeletedAppLocalStates() exec delete account app err: %w", err)
-			}
+			address := new(basics.Address)
+			*address = aa.Address
+
+			batch.Queue(deleteAccountAppStmtName, address[:], uint64(aa.App), uint64(round))
 		}
 	}
-
-	return nil
 }
 
-func (w *Writer) writeStateDelta(round basics.Round, delta ledgercore.StateDelta, specialAddresses transactions.SpecialAddresses) error {
-	err := w.writeAccountDeltas(round, delta.Accts, specialAddresses)
-	if err != nil {
-		return err
-	}
-
-	err = w.writeDeletedCreatables(round, delta.Creatables)
-	if err != nil {
-		return err
-	}
-
-	err = w.writeDeletedAssetHoldings(round, delta.ModifiedAssetHoldings)
-	if err != nil {
-		return err
-	}
-
-	err = w.writeDeletedAppLocalStates(round, delta.ModifiedAppLocalStates)
-	if err != nil {
-		return err
-	}
-
-	return nil
+func writeStateDelta(round basics.Round, delta ledgercore.StateDelta, specialAddresses transactions.SpecialAddresses, batch *pgx.Batch) {
+	writeAccountDeltas(round, delta.Accts, specialAddresses, batch)
+	writeDeletedCreatables(round, delta.Creatables, batch)
+	writeDeletedAssetHoldings(round, delta.ModifiedAssetHoldings, batch)
+	writeDeletedAppLocalStates(round, delta.ModifiedAppLocalStates, batch)
 }
 
-func (w *Writer) updateAccountSigType(payset []transactions.SignedTxnInBlock) error {
-	for _, stxnib := range payset {
-		if stxnib.Txn.RekeyTo == (basics.Address{}) {
-			sigtype, err := idb.SignatureType(&stxnib.SignedTxn)
+func updateAccountSigType(payset []transactions.SignedTxnInBlock, batch *pgx.Batch) error {
```

Comment:
what do you mean?

Reply:
Instead of setting the signature every time, we could set it only when there was a change.

Reply:
Suppose I'm rekeying an account. I'm supposed to set its sigtype to `nil`. However, I don't know if it's already `nil` because it was rekeyed immediately before or not `nil` because this account, being rekeyed already, made a non-rekey transaction immediately before.

Suppose I make a transaction that does not rekey. I'm supposed to set the sigtype of the account. But I don't know if the current sigtype in the DB is `nil` because the account was just rekeyed or not `nil` because there just was a non-rekey transaction.

Reply:
Ok, so in the worst case, we could set the sig type many times instead of just once per account. I guess it would require multiple passes over the data but it could be done.

I'm mainly noting that this is a spot where we could reduce the number of queries, no need to make the change now.